File: JavaScript/424. Longest Repeating Character Replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    var j=0;
    var m=0;
    var res=0;
    var map = {};
    for(var i=0;i<s.length;i++){
        if(map[s[i]] = map[s[i]]){
            map[s[i]]+=1;
        } 
        else{
            map[s[i]]=1;
        }
        m=Math.max(m,map[s[i]]);
        if(i-j+1-m>k){
            map[s[j]]--;
            j++;
        }
        res=Math.max(res, i-j+1);
    }
    return res;
};

File: Java/424. Longest Repeating Character Replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int j=0;
        int m=0;
        int res=0;
        HashMap<Character,Integer> map=new HashMap<>();
        for(int i=0;i<s.length();i++){
            map.put(s.charAt(i),map.getOrDefault(s.charAt(i),0)+1);
            m=Math.max(m,map.get(s.charAt(i)));
            if(i-j+1-m>k){
                map.put(s.charAt(j),map.get(s.charAt(j))-1);
                j++;
            }
            res=Math.max(res, i-j+1);
        }
        return res;
    }
}
